Source/NameServer/AddressBook/AddressBook.cpp

```diff
@@ -0,0 +1,1367 @@
+
+    return SUCCESS;
+}
+
+unsigned AddressBook::GetExternals(std::string &TaskName, const std::vector<Record_t>* &Records)
+{
+    TaskRecord_t *TRec = FindTask(TaskName);
+    if(TRec == PNULL)
+    {
+        return(ERR_TASK_NOT_FOUND);
+    }
+
+    if(!(TRec->TaskValid && TRec->MapValid && TRec->LinkValid))        // Return NotFound if dirty
+    {
+        return ERR_INVALID_MAP;
+    }
+
+    Records = &(TRec->Externals);
+
+    return SUCCESS;
+}
+
+unsigned AddressBook::GetSupervisors(std::string &TaskName, const std::vector<Record_t>* &Records)
+{
+    TaskRecord_t *TRec = FindTask(TaskName);
+    if(TRec == PNULL)
+    {
+        return(ERR_TASK_NOT_FOUND);
+    }
+
+    if(!(TRec->TaskValid && TRec->MapValid && TRec->LinkValid))        // Return NotFound if dirty
+    {
+        return ERR_INVALID_MAP;
+    }
+
+    Records = &(TRec->Supervisors);
+
+    return SUCCESS;
+}
+
+unsigned AddressBook::GetExtCon(std::string &TaskName, const RecordVect_t* &Records)
+{
+    TaskRecord_t *TRec = FindTask(TaskName);
+    if(TRec == PNULL)
+    {
+        return(ERR_TASK_NOT_FOUND);
+    }
+
+    if(!(TRec->TaskValid && TRec->MapValid && TRec->LinkValid))        // Return NotFound if dirty
+    {
+        return ERR_INVALID_MAP;
+    }
+
+    Records = &(TRec->ExtCon);
+
+    return SUCCESS;
+}
+
+
+
+
+//==============================================================================
+//Private Methods
+
+/*==============================================================================
+ * FindTask:    Return a pointer to the named task.
+ *============================================================================*/
+TaskRecord_t * AddressBook::FindTask(const std::string &TaskName)
+{
+    // Check that the task exists.
+    TaskMap_t::iterator TSearch = TaskMap.find(TaskName);
+    if (TSearch == TaskMap.end()) {
+        return(PNULL);
+    }
+
+    return TSearch->second;
+}
+
+
+unsigned AddressBook::MapDevice(TaskRecord_t *TRec, Record_t *DRec)
+{
+    // Add device to name map
+    TRec->NameMap.insert(NameMap_t::value_type(DRec->Name, DRec));
+
+    // Add the device to the correct DeviceType list.
+    TRec->DevTypes[DRec->DeviceType].second.push_back(DRec);
+
+    if(DRec->RecordType == DeviceExt)  // Device has an external conn
+    {
+        TRec->ExtCon.push_back(DRec);  // Add it to the list of such devices.
+    }
+
+    // Add the device to the correct attribute list
+    if(DRec->Attribute > -1)
+    {
+        TRec->AttrTypes[static_cast<unsigned>(DRec->Attribute)].second.push_back(DRec);
+    }
+
+    return SUCCESS;
+}
+
+unsigned AddressBook::LinkDevice(TaskRecord_t *TRec, Record_t *DRec)
+{
+    // Add Device to Address map.
+    TRec->AddrMap.insert(AddrMap_t::value_type(DRec->Address, DRec));
+
+    // Add the device to the correct Supervisor list.
+    // Supervisors may be loaded after devices, may need to build the map.
+    SupMap_t::iterator SSearch = TRec->SupMap.find(DRec->Supervisor);
```

Comment:
Just an observation: I get the impression that `TRec->SupMap` doesn't contain many entries (precisely one per box), so `find` is not going to be slow.